File: src/synthetic_vcf_generator/bed_parser.py
"""BED file parser for filtering VCF generation to specific genomic regions."""

from pathlib import Path
from typing import Dict, List, Tuple


class BEDIntervals:
    """Parser and storage for BED file intervals."""

    def __init__(self, bed_file_path: str):
        """
        Parse a BED file and store intervals by chromosome.

        Args:
            bed_file_path: Path to BED file (tab-separated: chr, start, end, ...)
        """
        self.intervals: Dict[str, List[Tuple[int, int]]] = {}
        self._parse_bed_file(bed_file_path)

    def _parse_bed_file(self, bed_file_path: str) -> None:
        """
        Parse BED file and store intervals.

        BED format: chr start end [name] [score] [strand] ...
        BED uses 0-based half-open coordinates [start, end)
        """
        bed_path = Path(bed_file_path)
        if not bed_path.exists():
            raise FileNotFoundError(f"BED file not found: {bed_file_path}")

        with open(bed_path, "r") as f:
            for line_num, line in enumerate(f, 1):
                line = line.strip()
                # Skip empty lines and comments
                if not line or line.startswith("#") or line.startswith("track") or line.startswith("browser"):
                    continue

                fields = line.split("\t")
                if len(fields) < 3:
                    raise ValueError(
                        f"Invalid BED format at line {line_num}: expected at least 3 fields, got {len(fields)}"
                    )

                chrom = fields[0]
                try:
                    start = int(fields[1])
                    end = int(fields[2])
                except ValueError as e:
                    raise ValueError(f"Invalid coordinates at line {line_num}: {e}")

                if start < 0 or end < 0:
                    raise ValueError(f"Negative coordinates at line {line_num}: start={start}, end={end}")
                if start >= end:
                    raise ValueError(f"Invalid interval at line {line_num}: start={start} >= end={end}")

                # Store interval for this chromosome
                if chrom not in self.intervals:
                    self.intervals[chrom] = []
                self.intervals[chrom].append((start, end))

        # Sort intervals for each chromosome by start position
        for chrom in self.intervals:
            self.intervals[chrom].sort()

    def get_chromosomes(self) -> List[str]:
        """Return list of chromosomes in BED file."""
        return list(self.intervals.keys())

    def get_intervals(self, chromosome: str) -> List[Tuple[int, int]]:
        """
        Get intervals for a specific chromosome.

        Args:
            chromosome: Chromosome name

        Returns:
            List of (start, end) tuples in BED format (0-based half-open)
        """
        return self.intervals.get(chromosome, [])

    def contains_position(self, chromosome: str, position: int) -> bool:
        """
        Check if a position (1-based VCF coordinate) is within any interval.

        Args:
            chromosome: Chromosome name
            position: 1-based position (VCF format)

        Returns:
            True if position is within any BED interval for this chromosome
        """
        if chromosome not in self.intervals:
            return False

        # Convert VCF 1-based position to BED 0-based coordinate
        bed_pos = position - 1

        # Binary search for efficiency
        intervals = self.intervals[chromosome]
        for start, end in intervals:
            if start <= bed_pos < end:
                return True
            if start > bed_pos:
                # Since sorted, no need to check further
                break

        return False

    def get_valid_positions(self, chromosome: str, positions: List[int]) -> List[int]:
        """
        Filter a list of positions to only those within BED intervals.

        Args:
            chromosome: Chromosome name
            positions: List of 1-based VCF positions

        Returns:
            Filtered list of positions that fall within BED intervals
        """
        if chromosome not in self.intervals:
            return []

        return [pos for pos in positions if self.contains_position(chromosome, pos)]

    def get_total_bases(self, chromosome: str = None) -> int:
        """
        Calculate total bases covered by intervals.

        Args:
            chromosome: If specified, only count bases for this chromosome.
                       If None, count all chromosomes.

        Returns:
            Total number of bases covered
        """
        if chromosome:
            intervals = self.intervals.get(chromosome, [])
            return sum(end - start for start, end in intervals)
        else:
            return sum(
                sum(end - start for start, end in intervals)
                for intervals in self.intervals.values()
            )

File: src/bed_parser.rs
use anyhow::{Context, Result};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use rand::Rng;

#[derive(Debug, Clone)]
pub struct BedIntervals {
    intervals: HashMap<String, Vec<(usize, usize)>>,
}

impl BedIntervals {
    pub fn new<P: AsRef<Path>>(bed_file_path: P) -> Result<Self> {
        let mut intervals: HashMap<String, Vec<(usize, usize)>> = HashMap::new();
        let content = fs::read_to_string(&bed_file_path)
            .with_context(|| format!("Failed to read BED file: {:?}", bed_file_path.as_ref()))?;

        for (line_num, line) in content.lines().enumerate() {
            let line = line.trim();
            // Skip empty lines and comments
            if line.is_empty()
                || line.starts_with('#')
                || line.starts_with("track")
                || line.starts_with("browser")
            {
                continue;
            }

            let fields: Vec<&str> = line.split('\t').collect();
            if fields.len() < 3 {
                anyhow::bail!(
                    "Invalid BED format at line {}: expected at least 3 fields, got {}",
                    line_num + 1,
                    fields.len()
                );
            }

            let chrom = fields[0].to_string();
            let start: usize = fields[1]
                .parse()
                .with_context(|| format!("Invalid start coordinate at line {}", line_num + 1))?;
            let end: usize = fields[2]
                .parse()
                .with_context(|| format!("Invalid end coordinate at line {}", line_num + 1))?;

            if start >= end {
                anyhow::bail!(
                    "Invalid interval at line {}: start={} >= end={}",
                    line_num + 1,
                    start,
                    end
                );
            }

            intervals.entry(chrom).or_insert_with(Vec::new).push((start, end));
        }

        // Sort intervals for each chromosome by start position
        for intervals_list in intervals.values_mut() {
            intervals_list.sort();
        }

        Ok(Self { intervals })
    }

    pub fn get_chromosomes(&self) -> Vec<String> {
        let mut chroms: Vec<String> = self.intervals.keys().cloned().collect();
        chroms.sort();
        chroms
    }

    pub fn get_intervals(&self, chromosome: &str) -> &[(usize, usize)] {
        self.intervals.get(chromosome).map(|v| v.as_slice()).unwrap_or(&[])
    }

    pub fn contains_position(&self, chromosome: &str, position: usize) -> bool {
        // Convert VCF 1-based position to BED 0-based coordinate
        let bed_pos = position.saturating_sub(1);

        if let Some(intervals) = self.intervals.get(chromosome) {
            for (start, end) in intervals {
                if *start <= bed_pos && bed_pos < *end {
                    return true;
                }
                if *start > bed_pos {
                    // Since sorted, no need to check further
                    break;
                }
            }
        }
        false
    }

    pub fn get_valid_positions(&self, chromosome: &str, positions: &[usize]) -> Vec<usize> {
        positions
            .iter()
            .filter(|pos| self.contains_position(chromosome, **pos))
            .copied()
            .collect()
    }

    pub fn get_total_bases(&self, chromosome: Option<&str>) -> usize {
        if let Some(chrom) = chromosome {
            self.intervals
                .get(chrom)
                .map(|intervals| {
                    intervals
                        .iter()
                        .map(|(start, end)| end - start)
                        .sum::<usize>()
                })
                .unwrap_or(0)
        } else {
            self.intervals
                .values()
                .map(|intervals| {
                    intervals
                        .iter()
                        .map(|(start, end)| end - start)
                        .sum::<usize>()
                })
                .sum()
        }
    }

    /// Sample positions directly from BED intervals using weighted sampling
    /// This is much more efficient than generate-then-filter approach
    pub fn sample_positions<R: rand::Rng>(
        &self,
        chromosome: &str,
        count: usize,
        rng: &mut R,
    ) -> Vec<usize> {
        use std::collections::HashSet;
        
        let intervals = match self.intervals.get(chromosome) {
            Some(ivs) if !ivs.is_empty() => ivs,
            _ => return Vec::new(),
        };

        // Calculate total bases in all intervals
        let total_bases: usize = intervals.iter().map(|(s, e)| e - s).sum();
        if total_bases == 0 {
            return Vec::new();
        }

        // Build cumulative weights for weighted sampling
        let mut cumulative_weights = Vec::with_capacity(intervals.len());
        let mut cumsum = 0usize;
        for (start, end) in intervals.iter() {
            cumsum += end - start;
            cumulative_weights.push(cumsum);
        }

        // Sample positions with deduplication
        let mut positions = HashSet::with_capacity(count);
        let max_attempts = count * 3; // Prevent infinite loops
        let mut attempts = 0;

        while positions.len() < count && attempts < max_attempts {
            attempts += 1;

            // Sample which interval to use (weighted by interval size)
            let weight = rng.gen_range(0..total_bases);
            let interval_idx = cumulative_weights
                .binary_search(&weight)
                .unwrap_or_else(|idx| idx);
            
            let (start, end) = intervals[interval_idx];
            
            // Sample position within the interval (convert to 1-based VCF coordinates)
            let pos = rng.gen_range(start..end) + 1;
            positions.insert(pos);
        }

        // Convert to sorted vector
        let mut result: Vec<usize> = positions.into_iter().collect();
        result.sort_unstable();
        result
    }

    /// Calculate how many variants should be generated for each chromosome
    /// based on their BED interval coverage
    pub fn distribute_variants_by_coverage(
        &self,
        total_variants: usize,
    ) -> HashMap<String, usize> {
        let total_bases = self.get_total_bases(None);
        if total_bases == 0 {
            return HashMap::new();
        }

        let mut distribution = HashMap::new();
        for chrom in self.get_chromosomes() {
            let chrom_bases = self.get_total_bases(Some(&chrom));
            let chrom_variants = ((total_variants as f64 * chrom_bases as f64) 
                / total_bases as f64).round() as usize;
            distribution.insert(chrom, chrom_variants);
        }
        
        distribution
    }
}

